### PR TITLE
citogo via lambda

### PR DESCRIPTION
**Pull request by heronhaye**

### CODE REVIEW

File: go/citogo/main.go [hunk 1]
```diff
@@ -148,106 +155,104 @@ func (r *runner) flushTestLogsToTemp(logName string, log bytes.Buffer) (string,
...
-
-	if outcome != fail || r.opts.Branch != "master" {
-		return
+	if outcome == OutcomeSuccess {
+		return nil
 	}
-	hook := os.Getenv("CITOGO_MASTER_FAIL_WEBHOOK")
-	if hook == "" {
-		return
+	if len(r.flakes) >= r.opts.Flakes {
```
Comment (heronhaye) [inline go/citogo/main.go]: I'm not sure exactly what the purpose of this config value is. Probably not using it right either.

Comment (maxtaco) [reply go/citogo/main.go]: it's the total number of flakes a test run is allowed before it just fails out

File: go/citogo/main.go [hunk 2]
```diff
@@ -116,6 +116,7 @@ func (r *runner) listTests() error {
 		return err
 	}
 	r.tests = filter(strings.Split(out.String(), "\n"))
+	sort.Strings(r.tests)
```
Comment (maxtaco) [inline go/citogo/main.go]: why?

File: go/citogo/main.go (hunk 1 above)
Comment (maxtaco) [reply go/citogo/main.go]: what about the test for `err == errTestFailed`?    Don't want to rerun the test if `runTestOnce` failed for a different reason

File: go/citogo/main.go [hunk 3]
```diff
@@ -148,106 +149,98 @@ func (r *runner) flushTestLogsToTemp(logName string, log bytes.Buffer) (string,
...
+	outcome2, _, err2 := r.runTestOnce(test, true /* isRerun */)
+	if err2 != nil {
+		return err2
 	}
-	return err
+	switch outcome2 {
+	case OutcomeFail:
+		return errTestFailed
+	case OutcomeSuccess:
+		fmt.Printf("FLK: %s\n", test)
```
Comment (maxtaco) [inline go/citogo/main.go]: don't need this...

File: go/citogo/main.go [hunk 4]
```diff
@@ -148,106 +149,98 @@ func (r *runner) flushTestLogsToTemp(logName string, log bytes.Buffer) (string,
...
 }
 
 var errTestFailed = errors.New("test failed")
 
-func (r *runner) runTestOnce(test string, canRerun bool, isRerun bool) (string, error) {
+// runTestOnce only returns an error if there was a problem with the test
+// harness code; it does not return an error if the test failed.
+func (r *runner) runTestOnce(test string, isRerun bool) (outcome Outcome, where string, err error) {
+	defer func() {
+		fmt.Printf("%s: %s %s\n", outcome.Abbrv(), test, where)
```
Comment (maxtaco) [inline go/citogo/main.go]: this now doesn't work as before. It used to report `FLK?` if you could rerun it and it failed. Please restore the previous behavior.

File: go/citogo/main.go (hunk 2 above)
Comment (heronhaye) [reply go/citogo/main.go]: For easier searching, but I reverted it

File: go/citogo/main.go (hunk 4 above)
Comment (heronhaye) [reply go/citogo/main.go]: Done

**Comment by heronhaye** [comment]
>what about the test for err == errTestFailed? Don't want to rerun the test if runTestOnce failed for a different reason

I changed it so runTestOnce only returns an error if it fails as a test harnesser, not if the test itself failed (which is reported in the outcome return var).

```
    outcome, where, err := r.runTestOnce(test, false /* isRerun */, canRerun)   
    if err != nil {                                                                
        return err                                                                 
    }                 
```

**Comment by heronhaye** [comment]
This is ready to look at again. I added some changes for the server as well.

File: go/citogo/s3.go [hunk 5]
```diff
@@ -64,33 +64,45 @@ func awsStringSign4(key string, date string, region string, service string, toSi
...
 	if err != nil {
 		return "", err
 	}
 	name += "-" + randSuffix + ".gz"
 
+	_, err = awsCall("s3", bucket+".s3.amazonaws.com", name, "PUT", buf)
+	if err != nil {
+		return "", err
+	}
+	where := fmt.Sprintf("(fetch with: `curl -s -o - https://%s.s3.amazonaws.com/%s | zcat -d`)", bucket, name)
```
Comment (maxtaco) [inline go/citogo/s3.go]: didn't you change this to make it easier to copy/paste?

Comment (heronhaye) [reply go/citogo/s3.go]: added back, not sure where it got lost

**Pull request merged**